### ISSUE TITLE
Skip Writing Equivalent attachedPictures

### ISSUE THREAD

**Issue by KrLite** [sbooth/SFBAudioEngine#378]
I noticed that whenever calling `audioFile.writeMetadata()` with non empty `attachedPictures` in the metadata, the saving time is about 30 seconds each, which is much longer than writing with empty `attachedPictures` (about 0.1 seconds each).

I know it's unavoidable because some pictures are large in data size, but as I'm making a metadata editor, I cannot expect users to edit these pictures arbitrarily, which means in most cases the pictures remain the same during `writeMetadata()` calls. However, the pictures will always be fully rewritten referring to the source code of audio files. So, it will be much better if equivalent pictures don't get rewritten every time.

Is it possible to skip writing equivalent pictures while saving? I think it can be done by comparing and extracting a set of changed pictures, instead of simply writing the given set of pictures.

**Comment by sbooth** [comment]
The metadata functions in `SFBAudioEngine` are thin wrappers around `TagLib`. I'll need to delve into taglib's source because I'm not sure if taglib behaves as you suggest or if it always rewrites everything. For certain formats it also probably depends if there is padding available for use. In any case, 30 seconds seems to be a long time for a file to save, but for a large file that has to be completely rewritten it's possible. Does it seem that certain formats are slower than others? How large are the attached pictures?

**Comment by KrLite** [comment]
> The metadata functions in `SFBAudioEngine` are thin wrappers around `TagLib`. I'll need to delve into taglib's source because I'm not sure if taglib behaves as you suggest or if it always rewrites everything. For certain formats it also probably depends if there is padding available for use. In any case, 30 seconds seems to be a long time for a file to save, but for a large file that has to be completely rewritten it's possible. Does it seem that certain formats are slower than others? How large are the attached pictures?

I've only tested pngs for now, and indeed, it's truly a large file. Its data representation is 7-8 digits in length. I may test more files later.

But besides the file size, I still believe it's very practical to compare the changes before writing these data, as image files are really not in the same scale as other strings and integers. So, I suppose comparing images before writing and omit useless overwrites can make a huge difference on the performance.

**Comment by KrLite** [comment]
Currently looking, `AudioFile` called taglib’s `setItem` to set cover tags, take mp4 for example:

https://github.com/sbooth/SFBAudioEngine/blob/main/Sources/CSFBAudioEngine/Metadata/SFBAudioMetadata+TagLibMP4Tag.mm#L271

As I’ve tested writing metadata with the SAME images and without any images, it apparently shows that taglib won’t handle item comparisons before writing tags. Writing with images of the same data consumes much more time than writing without any images.

So, I suppose it will be helpful to make special judgements when writing images, or even make a custom wrapper of `setItem` which automatically compares item changes and omit useless overwrites.

Hope it helps.

**Comment by sbooth** [comment]
I agree with your comments in principle . However, depending on the type of tag being written, the file format, and whether padding is available, avoiding overwrites isn't necessarily possible. I also think in general taglib always renders and writes the entire tag to the file regardless of changes (i.e. there is no change tracking).

This discussion might be better held on the taglib repo since it really is about taglib, but for MP4 specifically, `TagLib::MP4::File::save` calls `TagLib::MP4::Tag::save` which in turn renders the tag data to a byte vector which is then written to the file, taking advantage of any `free` atoms to avoid unnecessarily rewriting the entire file. There are no checks for duplicate data, but it's unclear to me whether that would actually make a difference because the addition of any atoms (such as a new comment `\251cmt`) could force the rewriting of the entire file if there is no available or insufficient padding.

This isn't to say that there aren't opportunities for optimization. If you want to share one of the large files I'll be happy to run a test with taglib alone to ensure that it isn't `AudioFile` causing the issue, or you could open an issue on the taglib repo to bring in the real taglib experts like @ufleisch.

**Comment by KrLite** [comment]
Interesting...

I tested some more files and found that it depends. 😕 When writing with literally nothing modified, it takes time (indicating the pictures are indeed overwritten, maybe), but not that much for every picture. The original picture is only `~86KB` in size, but it takes up 3 or 5 seconds to save than a `~8MB` picture, so very weird. In the end, the average writing speed is acceptable, ranging from 4 to 6 seconds each. I don't know why I got 30 seconds of waiting that day, and I cannot replicate a saving time beyond 10 seconds now. I have to admit that I was annoyed by that incidental outcome.
